cuda_core/cuda/core/style-guide.md

```diff
@@ -0,0 +1,1818 @@
+
+### 6. Other Public Classes and Functions
+
+Following the principal class or function, define other public classes and functions. These include:
+
+- **Auxiliary classes**: Supporting classes that are part of the public API (e.g., `DeviceMemoryResourceOptions` is an auxiliary class used by `DeviceMemoryResource`)
+- **Abstract base classes**: ABCs that define interfaces (e.g., `MemoryResource` in `_buffer.pyx`)
+- **Other public classes**: Additional classes exported by the module
+
+**All classes and functions in this section should be sorted alphabetically by name**, regardless of their relationship to the principal class. The principal class appears first as an exception to this rule.
+
+**Example:** In `_device_memory_resource.pyx`, `DeviceMemoryResource` is the principal class and appears first. Then `DeviceMemoryResourceOptions` appears after it (alphabetically after the principal class), even though it's an auxiliary/options class.
+
+### 7. Public Module Functions
+
+After all classes, define public module-level functions that are part of the API.
+
+### 8. Private or Implementation Functions
+
+Finally, define private functions and implementation details. These include:
+
+- Functions with names starting with `_` (private)
+- `cdef inline` functions used for internal implementation
+- Helper functions not part of the public API
+
+### Example Structure
+
+```python
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# Imports (cimports first, then regular imports)
+from libc.stdint cimport uintptr_t
+from cuda.core.experimental._memory._device_memory_resource cimport DeviceMemoryResource
+import abc
+
+__all__ = ['Buffer', 'MemoryResource', 'some_public_function']
+
+# Type aliases (if any)
+DevicePointerT = driver.CUdeviceptr | int | None
+"""Type union for device pointer representations."""
+
+# Principal class
+cdef class Buffer:
+    """Principal class for this module."""
+    # ...
+
+# Other public classes
+cdef class MemoryResource:
+    """Abstract base class."""
+    # ...
+
+# Public module functions
+def some_public_function():
+    """Public API function."""
+    # ...
+
+# Private implementation functions
+cdef inline void Buffer_close(Buffer self, stream):
+    """Private implementation helper."""
+    # ...
+```
+
+### Notes
+
+- Not every file will have all sections. For example, a utility module may not have a principal class.
+- The distinction between "principal" and "other" classes is based on the file's primary purpose. If a file exists primarily to define one class, that class is the principal class.
+- Private implementation functions should be placed at the end of the file to keep the public API visible at the top.
+- **Within each section**, classes and functions should be sorted alphabetically by name. The principal class or function is an exception to this rule, as it appears first in its respective section.
```

Comment:
I'm not sure this should be enforced without a linter. It feels a bit draconian to me to require someone to reorder their methods.

For example, I like to order methods by "relevance radius", which is a squishy concept that is something like "methods that are related to each other are close to each other". I find this makes debugging way easier, if only because I can see more related code in on viewpoint.